compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/SyntheticAccessorLowering.kt

```diff
@@ -167,7 +173,11 @@ internal class SyntheticAccessorLowering(val context: JvmBackendContext) : IrEle
     private fun IrDeclarationWithVisibility.accessorParent(parent: IrDeclarationParent = this.parent) =
         if (visibility == JavaVisibilities.PROTECTED_STATIC_VISIBILITY) {
             val classes = allScopes.map { it.irElement }.filterIsInstance<IrClass>()
-            classes.lastOrNull { parent is IrClass && it.isSubclassOf(parent) } ?: classes.last()
+            val companions = classes.mapNotNull { it.companionObject() }.filterIsInstance<IrClass>()
+            val objectsInScope =
+                classes.flatMap { it.declarations.filter(IrDeclaration::isAnonymousObject).filterIsInstance<IrClass>() }
+            val candidates = objectsInScope + companions + classes
+            candidates.lastOrNull { parent is IrClass && it.isSubclassOf(parent) } ?: classes.last()
```

Comment:
Feedback on these list gymnastics is more than welcome. I feel this could be done better.

Reply:
I think it's ok. I'm somewhat unsure about `?: classes.last()` clause, but if the tests like those in testData\codegen\box\objects\companionObjectAccess\multipleCompanionsWithAccessors pass, that's fine.

Reply:
Yeah, `classes.last()` is the fallback left in from previous version. You know, I never actually tried removing it, seeing what instances arent covered by the others.

I guess in principle the backend shouldn't error out?

Reply:
I actually prefer to error out in such cases, so that we identify problems fast. 
It _seems_ to me that `classes.last()` is ok because of how scopes are organized, although I'd better write some more tests with nested/inner classes.

Reply:
OK! Do you want me to do anything more for this PR?

Reply:
I think it's ok for now. Since the tests passed, I'll close the PR and push manually.